dash/util/s3_migration.py

```
@@ -0,0 +1,361 @@
+
+
+def setup_logging(level: str) -> None:
+    logging.basicConfig(
+        level=getattr(logging, level),
+        format="%(asctime)s | %(levelname)s | %(message)s",
+    )
+
+
+def parse_iso8601(ts: str) -> datetime:
+    """Parse a permissive ISO-8601 string into an aware UTC datetime.
+
+    Supports:
+      - "Z" suffix
+      - Offsets like "+07:00"
+      - Naive datetimes are assumed to be US/Pacific
+    """
+    s = ts.strip()
+    if s.endswith("Z"):
+        s = s[:-1] + "+00:00"
+    try:
+        dt = datetime.fromisoformat(s)
+    except ValueError as e:
+        raise SystemExit(f"Invalid ISO-8601 deadline: {ts}") from e
+    if dt.tzinfo is None:
+        dt = dt.replace(tzinfo=PACIFIC)
+    return dt.astimezone(timezone.utc)
+
+
+@dataclasses.dataclass
+class S3Object:
+    key: str
+    size: int
+    last_modified: datetime  # timezone-aware
+
+
+def list_objects(client, bucket: str, prefix: str) -> list[S3Object]:
+    paginator = client.get_paginator("list_objects_v2")
+    objects = []
+    for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
+        for item in page.get("Contents", []):
+            # Skip "directory" placeholders
+            key = item["Key"]
+            if key.endswith("/") and item.get("Size", 0) == 0:
+                continue
+            lm = item["LastModified"]
+            if lm.tzinfo is None:
+                lm = lm.replace(tzinfo=timezone.utc)
+            else:
+                lm = lm.astimezone(timezone.utc)
+            objects.append(
+                S3Object(key=key, size=item.get("Size", 0), last_modified=lm)
+            )
+    return objects
+
+
+def newest_last_modified(objs: Iterable[S3Object]) -> datetime | None:
+    newest = None
+    for obj in objs:
+        if newest is None or obj.last_modified > newest:
+            newest = obj.last_modified
+    return newest
+
+
+def relative_key_path(key: str, base_prefix: str) -> str:
+    assert key.startswith(base_prefix)
```

I generally don't like asserts since they can be disabled. Maybe raise a value error if the check fails instead?